aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -236,6 +236,56 @@ TORCH_PRECOMPUTE_META_FUNC(index_add)
   return TORCH_PRECOMPUTE_STRUCT(index_add)().set_dim(dim);
 }
 
+template <typename Meta>
+int64_t index_fill_meta_impl(
+    Meta& meta,
+    const Tensor& self,
+    int64_t dim,
+    const Tensor& index,
+    const Scalar& value) {
+
+  at::native::scatter_gather_dtype_check("index_fill", self, index);
+  at::assert_no_overlap(self, index);
+  if (at::has_internal_overlap(self) == at::MemOverlap::YES) {
+    TORCH_WARN(
+      "Use of index_fill on expanded tensors is deprecated. "
+      "Please clone() the tensor before performing this operation. "
+      "This also applies to advanced indexing e.g. tensor[mask] = scalar");
+  }
+  if (!self.is_complex() && value.isComplex()) {
+    TORCH_CHECK(false, "index_fill(): Converting complex Scalar to non-complex type is not supported");
+  }
```

Comment:
This should probably by a `canCast` to be more general.
```suggestion
    TORCH_CHECK(canCast(self.scalar_type(), value.type()), "index_fill(): Converting ",
    value.type(), " to ",  self.scalar_type()," is not supported");
```

Reply:
This almost works except the case of casting value (long or float) to bool input tensor.
Master version does the following today:
```python
import torch
print(torch.__version__)  # 1.10.0+cu111

dtype = torch.bool
device = "cpu"
x = torch.tensor([[0, 1], [2, 3]], dtype=dtype, device=device)
index = torch.tensor([1], device=device)
x.index_fill_(1, index, 3)   # value is long
> tensor([[False,  True],
                [False,  True]])
```
or 
```python
x = torch.tensor([[0, 1], [2, 3]], dtype=dtype, device=device)
index = torch.tensor([1], device=device)
x.index_fill_(1, index, 3.0)   # value is float
> tensor([[False,  True],
                [False,  True]])
```

Using `canCast` we'll raise : `RuntimeError: index_fill(): Converting Long to Bool is not supported`
Do we want this change ? @mruberry what do you think ?